hypothesis-python/src/hypothesis/strategies/_internal/utils.py

```diff
@@ -196,3 +190,22 @@ def to_jsonable(obj: object, *, avoid_realization: bool) -> object:
 
     # If all else fails, we'll just pretty-print as a string.
     return pretty(obj)
+
+
+def to_jsonable(obj: object, *, avoid_realization: bool) -> object:
+    """Recursively convert an object to json-encodable form.
+
+    This is not intended to round-trip, but rather provide an analysis-ready
+    format for observability.  To avoid side affects, we pretty-print all but
+    known types.
+    """
+
+    # lift out a helper method to handle recursion errors at the top level.
+    # In an ideal world, _to_jsonable would detect reference cycles itself,
+    # in the same way that pretty.pretty does, so we get the nice hypothesis-specific
+    # logic of to_jsonable, and also to avoid the performance penalty of recursing
+    # a thousand times before falling back.
```

Comment:
Well, why not add reference-cycle detection?  `seen: WeakValueDict[int, object] | None` with keys `id(obj)` should make this fairly easy?  (and then, yeah, I think falling back to `pretty(obj)` on recursion is good)